sfge/src/main/java/com/salesforce/telemetry/TelemetryUtil.java

```diff
@@ -0,0 +1,101 @@
+package com.salesforce.telemetry;
+
+import com.google.gson.Gson;
+import com.salesforce.messaging.CliMessager;
+import com.salesforce.messaging.EventKey;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+public final class TelemetryUtil {
+    private static final String MAIN_THREAD_NAME = "main";
+
+    /**
+     * Posts a telemetry event in response to a {@link org.apache.logging.log4j.Logger#warn} call.
+     *
+     * @param message - The message being logged.
+     */
+    public static void postWarningTelemetry(String message) {
+        EventType eventType =
+                isMainThread() ? EventType.MAIN_THREAD_WARNING : EventType.RULE_THREAD_WARNING;
+        postTelemetry(message, eventType);
+    }
+
+    /**
+     * Posts a telemetry event in response to an {@link
+     * com.salesforce.exception.SfgeRuntimeException} being thrown.
+     *
+     * @param message - The exception's message, if available.
+     */
+    public static void postExceptionTelemetry(String message) {
+        EventType eventType =
+                isMainThread() ? EventType.MAIN_THREAD_EXCEPTION : EventType.RULE_THREAD_EXCEPTION;
+        postTelemetry(message, eventType);
+    }
+
+    private static void postTelemetry(String message, EventType eventType) {
+        TelemetryData telemetryData = new TelemetryData(message, eventType);
+        CliMessager.postMessage(
+                "TelemetryData", EventKey.INFO_TELEMETRY, new Gson().toJson(telemetryData));
+    }
+
+    /** Returns a boolean indicating whether this thread is the main thread. */
+    private static boolean isMainThread() {
+        // Assumption: The name of the main thread will consistently be "main".
+        // If this assumption is invalidated, this code must change.
+        return MAIN_THREAD_NAME.equalsIgnoreCase(Thread.currentThread().getName());
```

Comment:
This turned out to be a far easier way of verifying whether the current thread is `main`, and I couldn't find any information about the main thread not being named `main` ever.

Reply:
Looked around and can't find another name unless the developer decides to rename it. Since the code is ours, we might be fine with this. Agree that we shouldn't spend more cycles on this.